src/FirebasePushSender.php

```diff
@@ -0,0 +1,123 @@
+ */
+
+namespace Askvortsov\FlarumPWA;
+
+use Flarum\Discussion\Discussion;
+use Flarum\Notification\Blueprint\BlueprintInterface;
+use Flarum\Notification\MailableInterface;
+use Flarum\Post\CommentPost;
+use Flarum\Post\Post;
+use Kreait\Firebase\Contract\Messaging;
+use Kreait\Firebase\Messaging\CloudMessage;
+use Kreait\Firebase\Messaging\Notification;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+class FirebasePushSender
+{
+    private Messaging $messaging;
+
+    protected TranslatorInterface $translator;
+
+    public function __construct(
+        Messaging $messaging,
+        TranslatorInterface $translator,
+    ) {
+        $this->messaging = $messaging;
+        $this->translator = $translator;
+    }
+
+    /**
+     * @throws \ReflectionException
+     */
+    public static function canSend(string $blueprintClass): bool
+    {
+        return (new \ReflectionClass($blueprintClass))->implementsInterface(MailableInterface::class) || in_array(
+            $blueprintClass,
+            static::$SUPPORTED_NON_EMAIL_BLUEPRINTS
+        );
+    }
+
+    public function notify(BlueprintInterface $blueprint, array $userIds = []): void
+    {
+        FirebasePushSubscription::whereIn('user_id', $userIds)->each(function (FirebasePushSubscription $subscription) use ($blueprint) {
+            $this->messaging->send(
+                $this->newMessage($subscription, $blueprint)
+            );
+        });
+    }
+
+    private function newMessage(FirebasePushSubscription $subscription, BlueprintInterface $blueprint): CloudMessage
+    {
+        return CloudMessage::new()
+            ->withTarget('token', $subscription->token)
+            ->withNotification(
+                Notification::fromArray([
+                    'title' => $this->getTitle($blueprint),
+                    'body' => strip_tags($this->getBody($blueprint)),
+                ])
+            );
+    }
+
+    private function getBody(BlueprintInterface $blueprint)
+    {
+        $content = '';
+
+        $subject = $blueprint->getSubject();
+
+        switch ($blueprint::getSubjectModel()) {
+            case Discussion::class:
+                /** @var Discussion $subject */
+                $content = $this->getRelevantPostContent($subject);
+                break;
+            case Post::class:
+                /** @var Post $subject */
+                if ($subject instanceof CommentPost) {
+                    $content = $subject->formatContent();
+                }
+                break;
+        }
+
+        return $content;
+    }
+
+    protected function getRelevantPostContent($discussion): string
+    {
+        $relevantPost = $discussion->mostRelevantPost ?: $discussion->firstPost ?: $discussion->comments->first();
+
+        if ($relevantPost === null) {
+            return '';
+        }
+
+        return $relevantPost->formatContent();
+    }
+
+    protected function getTitle(BlueprintInterface $blueprint): string
```

Comment:
Is there a reason we duplicate this vs splitting some common helpers out from the existing push implementation?